src/dl_node.cpp

```diff
@@ -30,132 +31,36 @@ namespace ovms {
 
 const uint WAIT_FOR_STREAM_ID_TIMEOUT_MICROSECONDS = 1;
 
-Status DLNode::execute(ThreadSafeQueue<std::reference_wrapper<Node>>& notifyEndQueue) {
-    Status status;
-    if (this->nodeStreamIdGuard == nullptr) {
-        status = requestExecuteRequiredResources();
-        if (!status.ok()) {
-            notifyEndQueue.push(*this);
-            return status;
-        }
-    }
-    auto streamId = this->nodeStreamIdGuard->tryGetId(WAIT_FOR_STREAM_ID_TIMEOUT_MICROSECONDS);
-    if (!streamId) {
-        SPDLOG_DEBUG("[Node: {}] Could not acquire stream Id right away", getName());
-        return StatusCode::PIPELINE_STREAM_ID_NOT_READY_YET;
-    }
-    auto& inferRequestsQueue = this->model->getInferRequestsQueue();
-    auto& inferRequest = inferRequestsQueue.getInferRequest(streamId.value());
-    status = setInputsForInference(inferRequest);
-    if (!status.ok()) {
-        notifyEndQueue.push(*this);
-        return status;
-    }
-    status = executeInference(notifyEndQueue, inferRequest);
-    if (!status.ok()) {
-        notifyEndQueue.push(*this);
-        return status;
-    }
-    return status;
-}
-
-Status DLNode::requestExecuteRequiredResources() {
-    Status status = StatusCode::OK;
-    status = this->modelManager.getModelInstance(
-        this->modelName,
-        this->modelVersion.value_or(0),
-        this->model,
-        this->modelUnloadGuard);
-
-    if (!status.ok()) {
-        SPDLOG_DEBUG("Getting modelInstance failed for node: {} with: {}", getName(), status.string());
-        return status;
-    }
-
-    status = prepareInputsAndModelForInference();
-    if (!status.ok()) {
-        return status;
-    }
-    auto& inferRequestsQueue = this->model->getInferRequestsQueue();
-    this->nodeStreamIdGuard = std::make_unique<NodeStreamIdGuard>(inferRequestsQueue);
-    return status;
+Status DLNode::execute(session_key_t sessionKey, PipelineEventQueue& notifyEndQueue) {
+    auto& nodeSession = getNodeSession(sessionKey);                   // TODO error check?
+    auto& dlNodeSession = dynamic_cast<DLNodeSession&>(nodeSession);  // TODO error check dynamic_cast
+    return dlNodeSession.execute(notifyEndQueue, WAIT_FOR_STREAM_ID_TIMEOUT_MICROSECONDS, *this);
```

Comment:
Why not to have `NodeSession` as base class and let `DLNodeSession` inherit and implement separate _execute_ method. This way we would not needed to dynamic cast node session. (Similarly to `DLNode->Node`)

Reply:
Ok I see now you did inherit from `NodeSession`. Why not to have _execute_ implementation for all inherited classes?

Reply:
Beacause the parameters needed for execute for different classes differ significantly. It would mean that nodesession would have to know that something like InferRequest/ModelInstance exist.

What I had doubts as well was how to divide responsibility over DLNodeSession & DLNode since first does contain all required resources for execution while the second does have the knowledge about DAG topology. I did what seemed the most reasonable but I might got tunnel vision at some point.